perceval/utils/qmath.py

```diff
@@ -61,3 +64,48 @@ def exponentiation_by_squaring(base, power: int):
         temp_base = temp_base * temp_base
 
     return result
+
+
+def kmeans(features: np.ndarray, n_clusters: int, n_init: int = 10) -> np.ndarray:
+    """
+    Manual KMeans implementation.
+
+    :param features: Data points for clustering.
+    :type features: np.ndarray
+    :param n_clusters: Number of clusters.
+    :type n_clusters: int
+    :param n_init: Number of times the k-means algorithm will be run with different centroid seeds.
+    :type n_init: int
+    :return: Cluster labels.
+    :rtype: np.ndarray
+    """
+    best_labels = None
+    best_inertia = float('inf')
+
+    for _ in range(n_init):
+        # Initialize centroids randomly
+        indices = np.random.choice(features.shape[0], n_clusters, replace=False)
+        centroids = features[indices]
+
+        for _ in range(300):  # Maximum number of iterations
```

Comment:
This kind of hard-coded values can cause bugs. The usual way we implement them is by `VARIABLE_NAME=300`, then use `VARIABLE_NAME` in place of the hard coded value